rfcs/libraries.md

```diff
@@ -0,0 +1,121 @@
+# Libraries
+
+## Summary
+
+We need a way to group together related code and provide easy ways to require them.
+
+## Motivation
+
+- Package management
+- Third party libraries
+- Simpler requires
+
+## Design
+
+### String Syntax
+
+Our require syntax should:
+
+- Work the same way in the Roblox engine and Luau language
+- Be resolvable at compile-time
+
+#### Absolute
+
+Modules can be required by their absolute path from the root of the filesystem.
+
+Requiring a file called `MyModule.luau` in `C:/MyLibrary`:
+```lua=
+local MyModule = require("C:/MyLibrary/MyModule")
```

Comment:
I am assuming strings passed to require must be string literals.  (Maybe we should call that out in the RFC just to be more specific.)

Reply:
If that is the case, how would one semi-dynamically require all the modules (lua files) in a folder? That seems like a reasonable use-case this should support.

Reply:
Any string can be used, it does not have to be a string literal. For type-checking etc. though I expect a string literal to be necessary.

Reply:
The RFC says that require should "Be resolvable at compile-time".  So it seems like either we need to restrict ourselves to string literals, or strings that can be determined at compile time (i.e. string constant expressions).
Or, maybe we just need to remove that requirement if we don't need require to be resolved at compile time (although that could be problematic for type checking or static analysis, so I am not really suggesting that).

Reply:
Yeah I think the issue is the phrasing in "be resolvable at compile-time." We should remove that. It makes it sound like the program is ill-formed and won't execute if we have something like `local M = require(JoinPath("$MyLib", "Foo"))`.

Or we can change the phrasing so that it says something like: "Luau type checker may require paths to be resolvable statically to guarantee type safety across modules."

Reply:
I should probably change it to say

> String literals must be resolvable at compile-time, but all strings should be supported

Reply:
I have clarified this as:

>  Support compile-time file resolution _where possible_ for type checking.